Giga postprocessing

kapodamy opened the pull request:
This is a rewrite of most of giga

this PR include:
* remove interfaces with one implementation
* fix download resources with unknow length
* marquee style for ProgressDrawable
* "view details" option in mission context menu
* notification for finished downloads
* postprocessing infrastructure: sub-missions, circular file, layers for layers of abstractions for Java IO streams
* Mp4 muxing (only DASH brand)
* WebM muxing
* Captions downloading
* alert dialog for overwrite existing downloads finished or not.

Notes:
* first commit bring lastest dev branch
* unfinished downloads will be erased (new serialVersionUID)
* dowload database version bumped to 3
* DeleteDownloadManager is broken
* CircularFile is like a circular buffer but in a file. This prevent use extra storage space


- [✔] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.

theScrabi commented:
@kapodamy great work, but Aaaaa I need help reviewing code O.o

kapodamy commented:
@theScrabi requires https://github.com/TeamNewPipe/NewPipeExtractor/pull/118

theScrabi commented on app/src/main/java/org/schabi/newpipe/util/Mp4DashReader.java:
```diff
@@ -0,0 +1,911 @@
+                tracks[i].kind = TrackKind.Video;
+            }
+
+            /*switch (stat) {
```
Unused code

theScrabi commented on app/src/main/java/org/schabi/newpipe/util/Mp4DashReader.java:
```diff
@@ -0,0 +1,911 @@
+                return obj;
+            }
+
+            /*Traf tmp = parse_traf(box, trackId);
```
unused code

theScrabi commented on app/src/main/java/org/schabi/newpipe/util/Mp4DashReader.java:
```diff
@@ -0,0 +1,911 @@
+        traf.trun = parse_trun();
+        ensure(b);
+
+        /*if (traf.trun.entryCount > 0) {
```
unused code

theScrabi commented on app/src/main/java/org/schabi/newpipe/util/Mp4DashWriter.java:
```diff
@@ -0,0 +1,622 @@
+package org.schabi.newpipe.util;
+
+import us.shandian.giga.postprocessing.io.SharpStream;
```
gigaget should not be used from newpipe side.

theScrabi commented on app/src/main/java/org/schabi/newpipe/util/Mp4DashWriter.java:
```diff
@@ -0,0 +1,622 @@
+ *
+ * @author kapodamy
+ */
+public class Mp4DashWriter {
```
Class belongs into extractor.

theScrabi commented on app/src/main/java/org/schabi/newpipe/util/SubtitleConverter.java:
```diff
@@ -0,0 +1,321 @@
+import java.text.ParseException;
+import java.util.Locale;
+
+import us.shandian.giga.postprocessing.io.CircularFile;
```
This should not be used.

theScrabi commented on app/src/main/java/org/schabi/newpipe/util/SubtitleConverter.java:
```diff
@@ -0,0 +1,321 @@
+
+import us.shandian.giga.postprocessing.io.CircularFile;
+
+public class SubtitleConverter {
```
Class belongs into backend.

theScrabi commented on app/src/main/java/org/schabi/newpipe/util/WebMReader.java:
```diff
@@ -0,0 +1,511 @@
+ *
+ * @author kapodamy
+ */
+public class WebMReader {
```
This class should also be part of extractor.

theScrabi commented on app/src/main/java/org/schabi/newpipe/util/WebMWriter.java:
```diff
@@ -0,0 +1,728 @@
+ *
+ * @author kapodamy
+ */
+public class WebMWriter {
```
This class should be part of extractor.

theScrabi commented on app/src/main/java/org/schabi/newpipe/util/XmlDocument.java:
```diff
@@ -0,0 +1,208 @@
+import java.util.HashMap;
+import java.util.Map;
+
+class XmlDocument {
```
Why do we have to implement something like this ourself, can't we use libraries for xml parsing?

kapodamy replied:
perhaps

theScrabi commented on app/src/main/java/org/schabi/newpipe/util/XmlDocument.java:
```diff
@@ -0,0 +1,208 @@
+        fac = factory;
+    }
+
+    XmlNode SelectSingleNode(String... path) throws XmlPullParserException, IOException {
```
c# styled function name

theScrabi commented on app/src/main/java/org/schabi/newpipe/util/XmlDocument.java:
```diff
@@ -0,0 +1,208 @@
+        return new XmlNode(parser);
+    }
+
+    XmlNodeList SelectNodes(String... path) throws XmlPullParserException, IOException {
```
c# styled function name

theScrabi commented on app/src/main/java/org/schabi/newpipe/util/XmlDocument.java:
```diff
@@ -0,0 +1,208 @@
+    }
+}
+
+class XmlNode {
```
Two classes in one file is not permitted acording to java, or indent is wrong.

theScrabi commented on app/src/main/java/org/schabi/newpipe/util/XmlDocument.java:
```diff
@@ -0,0 +1,208 @@
+    private Map<String, String> attrs;
+}
+
+class XmlNodeList {
```
Two or more classes in one file are not permitted in java.

kapodamy commented:
now https://github.com/TeamNewPipe/NewPipeExtractor/pull/119

kapodamy commented:
Fixed: #1762 #1655 
Maybe fixed: #1325 #602
Implemented: #1460 #1408 #1338 (no workaround) #1272 (thread count only) #961 #457 #355

theScrabi commented:
@kapodamy can you please check why the build is failing.

kapodamy commented:
@theScrabi jitpack is providing an empty jar [https://jitpack.io/#TeamNewPipe/NewPipeExtractor/91b1efc97e](https://jitpack.io/#TeamNewPipe/NewPipeExtractor/91b1efc97e).
and only contains a manifest, see [
see https://jitpack.io/com/github/TeamNewPipe/NewPipeExtractor/91b1efc97e/NewPipeExtractor-91b1efc97e.jar](https://jitpack.io/com/github/TeamNewPipe/NewPipeExtractor/91b1efc97e/NewPipeExtractor-91b1efc97e.jar)

TobiGr commented:
It is great to see this PR compiling again :) I did a small test and I have to admit that I love your changes. Let's see what we can improve to make this PR even better.


When a download is finished the notification informs the user about this. I think to make clear which app set up this notification, we should add NewPipe's logo to it. Additionally, we could enable line breaks whenever the title is too long.
![download_finished_notification_icon](https://user-images.githubusercontent.com/17365767/48733442-01164200-ec43-11e8-837c-e620a823378d.png)

Using the light theme, the titles in the downloads list are hard to read.
![download_queue_title_ligh_theme](https://user-images.githubusercontent.com/17365767/48733444-01aed880-ec43-11e8-8cc1-0c613d843873.png)

After queuing a video, it shows an estimated size of 2MB which is clearly incorrect. Is it possible to pass the file size we fetched when showing the download dialog? 
![download_queue_estimated_file_size](https://user-images.githubusercontent.com/17365767/48733523-2e62f000-ec43-11e8-9c4b-9ae5fe6e77d7.png)

Riotous commented:
Finally, the PR so many of us have been waiting for!

TobiGr commented:
I started 2 downloads and tested the queuing behaviour (started and stopped the downloads twice). Everything was fine and the downloads continued. After a while the screen went off and I did not pay attention to the downloads anymore. After two or three minutes I realized the downloads stopped with following errors:

![download_error_2](https://user-images.githubusercontent.com/17365767/48797612-f622e680-ed02-11e8-8261-b8fe8e85f592.png)
![download_error_1](https://user-images.githubusercontent.com/17365767/48797611-f622e680-ed02-11e8-965a-02266068b72e.png)


Unfortunately, I was not able to restart the download.

--------------------------


The downloads view crashes when rotating the screen.


## Exception
* __User Action:__ ui error
* __Request:__ App crash, UI failure
* __Content Language:__ GB
* __Service:__ none
* __Version:__ 0.14.2
* __OS:__ Linux Android 5.0.1 - 21


<details><summary><b>Crash log</b></summary><p>

```
java.lang.NullPointerException: Attempt to invoke interface method 'android.view.MenuItem android.view.MenuItem.setVisible(boolean)' on a null object reference
	at us.shandian.giga.ui.adapter.MissionAdapter.onBindViewHolder(MissionAdapter.java:145)
	at android.support.v7.widget.RecyclerView$Adapter.onBindViewHolder(RecyclerView.java:6781)
	at android.support.v7.widget.RecyclerView$Adapter.bindViewHolder(RecyclerView.java:6823)
	at android.support.v7.widget.RecyclerView$Recycler.tryBindViewHolderByDeadline(RecyclerView.java:5752)
	at android.support.v7.widget.RecyclerView$Recycler.tryGetViewHolderForPositionByDeadline(RecyclerView.java:6019)
	at android.support.v7.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:5858)
	at android.support.v7.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:5854)
	at android.support.v7.widget.LinearLayoutManager$LayoutState.next(LinearLayoutManager.java:2230)
	at android.support.v7.widget.GridLayoutManager.layoutChunk(GridLayoutManager.java:557)
	at android.support.v7.widget.LinearLayoutManager.fill(LinearLayoutManager.java:1517)
	at android.support.v7.widget.LinearLayoutManager.onLayoutChildren(LinearLayoutManager.java:612)
	at android.support.v7.widget.GridLayoutManager.onLayoutChildren(GridLayoutManager.java:171)
	at android.support.v7.widget.RecyclerView.dispatchLayoutStep2(RecyclerView.java:3924)
	at android.support.v7.widget.RecyclerView.dispatchLayout(RecyclerView.java:3641)
	at android.support.v7.widget.RecyclerView.onLayout(RecyclerView.java:4194)
	at android.view.View.layout(View.java:15905)
	at android.view.ViewGroup.layout(ViewGroup.java:5108)
	at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1959)
	at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1813)
	at android.widget.LinearLayout.onLayout(LinearLayout.java:1722)
	at android.view.View.layout(View.java:15905)
	at android.view.ViewGroup.layout(ViewGroup.java:5108)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:633)
	at android.widget.FrameLayout.onLayout(FrameLayout.java:568)
	at android.view.View.layout(View.java:15905)
	at android.view.ViewGroup.layout(ViewGroup.java:5108)
	at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1959)
	at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1813)
	at android.widget.LinearLayout.onLayout(LinearLayout.java:1722)
	at android.view.View.layout(View.java:15905)
	at android.view.ViewGroup.layout(ViewGroup.java:5108)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:633)
	at android.widget.FrameLayout.onLayout(FrameLayout.java:568)
	at android.view.View.layout(View.java:15905)
	at android.view.ViewGroup.layout(ViewGroup.java:5108)
	at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1959)
	at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1813)
	at android.widget.LinearLayout.onLayout(LinearLayout.java:1722)
	at android.view.View.layout(View.java:15905)
	at android.view.ViewGroup.layout(ViewGroup.java:5108)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:633)
	at android.widget.FrameLayout.onLayout(FrameLayout.java:568)
	at android.view.View.layout(View.java:15905)
	at android.view.ViewGroup.layout(ViewGroup.java:5108)
	at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1959)
	at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1813)
	at android.widget.LinearLayout.onLayout(LinearLayout.java:1722)
	at android.view.View.layout(View.java:15905)
	at android.view.ViewGroup.layout(ViewGroup.java:5108)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:633)
	at android.widget.FrameLayout.onLayout(FrameLayout.java:568)
	at android.view.View.layout(View.java:15905)
	at android.view.ViewGroup.layout(ViewGroup.java:5108)
	at android.view.ViewRootImpl.performLayout(ViewRootImpl.java:2425)
	at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2131)
	at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1256)
	at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:6443)
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:800)
	at android.view.Choreographer.doCallbacks(Choreographer.java:603)
	at android.view.Choreographer.doFrame(Choreographer.java:572)
	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:786)
	at android.os.Handler.handleCallback(Handler.java:815)
	at android.os.Handler.dispatchMessage(Handler.java:104)
	at android.os.Looper.loop(Looper.java:194)
	at android.app.ActivityThread.main(ActivityThread.java:5576)
	at java.lang.reflect.Method.invoke(Native Method)
	at java.lang.reflect.Method.invoke(Method.java:372)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:956)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:751)

```
</p></details>

kapodamy commented:
@TobiGr 
- implemented
- implemented
- re-implemented, but works only if the required streams sizes are fetched (if it is displayed in the download dialog). In weak networks maybe no.
- about `Show info`, this also happens when navigating between suggested videos. can't fix this

kapodamy commented:
@TobiGr about the errors
1.It seems that NewPipe dies when the screen goes off. Once the postprocessing is started, it can not be stopped or the file will be damaged (error 1007).
2. looks like a race-condition, should be fixed now

TobiGr commented:
@kapodamy It good to see that you were able to fix the crashes. Unfortunately, I produced some more :)
Just to note, I reviewed this at the state of 767dc2043288f72216f83347e87df925bdbfb712 (Thursday morning). So in case you fixed any of the following, don't be mad with me :) 
Let's start with small things and UX and bugs at last:

- [ ] In the settings you added the option to set the maximum number of retries when a download failed. When changing the number via the seek bar, the number on the right is not updated simultaneously.

![download_settings_retries](https://user-images.githubusercontent.com/17365767/48958815-dd643c00-ef61-11e8-88c2-83a276800a11.png)

- [ ] There are some more issues with the light theme which I didn't recognize before. 

![download_app_bar_theme](https://user-images.githubusercontent.com/17365767/48958814-dd643c00-ef61-11e8-8372-2a60a42bb6a3.png)

- [ ] There's also another small theme bug, but I guess we can open a separate ticket for it. After opening the downloads view, go to the settings via the 3-dot-menu in the action bar. Change a theme and go back to downloads via the back buttons. You'll see that the new theme was not applied. I didn't check it, but I guess this issue existed in previous versions, too.

- [ ] You can get the following exception when taping on a tile of a finished download. This crash does not appear when using the tile menu and then pressing play. I am not sure if this crash was there before. Nevertheless we are not able to "play" subtitles. So neither tapping on it or using the play option via the menu should be possible.
#### Exception
* __User Action:__ ui error
* __Request:__ App crash, UI failure
* __Content Language:__ GB
* __Service:__ none
* __Version:__ 0.14.2
* __OS:__ Linux Android 5.0.1 - 21


<details><summary><b>Crash log</b></summary><p>

```
java.lang.ClassCastException: us.shandian.giga.get.FinishedMission cannot be cast to us.shandian.giga.get.DownloadMission
	at us.shandian.giga.ui.adapter.MissionAdapter$ViewHolderItem.lambda$new$1(MissionAdapter.java:591)
	at us.shandian.giga.ui.adapter.-$$Lambda$MissionAdapter$ViewHolderItem$FHK8sq2xsmJ5bNl5JAWgRX95u8I.onClick(lambda)
	at android.view.View.performClick(View.java:4807)
	at android.view.View$PerformClick.run(View.java:20106)
	at android.os.Handler.handleCallback(Handler.java:815)
	at android.os.Handler.dispatchMessage(Handler.java:104)
	at android.os.Looper.loop(Looper.java:194)
	at android.app.ActivityThread.main(ActivityThread.java:5576)
	at java.lang.reflect.Method.invoke(Native Method)
	at java.lang.reflect.Method.invoke(Method.java:372)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:956)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:751)

```
</p></details>
<hr>


### Download bugs

There are issues with large downloads on Android 5. I experienced strange behaviours with streams >= 3 GB. As you can see, the download state is not displayed correctly and the current progress is not displayed either. Additionally, the download does not finish and at some point the download size (on the bottom left) grows every second. When opening the downloads view, for half a second or so, the current file site is displayed at the position of the current download speed in GB/s. I guess all these bugs are related to the incorrect state. I tested this with Android 5 and bad WiFi. Both are not optimal testing conditions, so in case you cannot reproduce this, I can try to create some more logs for you.


![download_wrong_state](https://user-images.githubusercontent.com/17365767/48958812-dccba580-ef61-11e8-91dc-68422af54689.png)
![download_wrong_stream size](https://user-images.githubusercontent.com/17365767/48958813-dccba580-ef61-11e8-8175-e68eb153bf32.png)

----

I'll take a look at your latest changes, which I didn't test, next week.

kapodamy commented:
with the latest changes, the application will crash (only in the debug version), because I removed the search bar from the "download settings" and replaced it with a list.

kapodamy commented:
@theScrabi
when you have time, please take a look

theScrabi commented:
The only issue is that the icons in the download manager do not adapt bright design. Pleas fix this in an additional commit.

![screenshot_20181211-163029](https://user-images.githubusercontent.com/1891273/49811314-4b32a500-fd63-11e8-9e9a-93100a55236e.png)

FaizKhan5 commented:
is it able to download audio in mp3 ?

gkeegan commented:
I don't believe so from what I've seen in this thread.

kapodamy commented:
@FaizKhan5 if you mean convert m4a to mp3 no.

FaizKhan5 commented:
@kapodamy yes, i mean that. Is there any chance to implement that in near future ?

pro4tlzz commented:
Thanks for the update
